api/etcd/get.go

```diff
@@ -49,28 +49,34 @@ func (s *RulesetService) Get(ctx context.Context, path, version string) (*regula
 		return nil, errors.Wrap(err, "failed to unmarshal signature")
 	}
 
-	// decode versions
-	var versions pb.Versions
-	err = proto.Unmarshal(resp.Responses[1].GetResponseRange().Kvs[0].Value, &versions)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to unmarshal versions")
+	// the versions must be parsed from the keys returned by the second operation
+	versions := make([]string, len(resp.Responses[1].GetResponseRange().Kvs))
+	for i, kv := range resp.Responses[1].GetResponseRange().Kvs {
+		_, versions[i] = s.pathVersionFromKey(string(kv.Key))
 	}
 
-	// decode rules
+	// decode rules, might not be filled if only the signature was created
 	var rules pb.Rules
-	err = proto.Unmarshal(resp.Responses[2].GetResponseRange().Kvs[0].Value, &rules)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to unmarshal ruleset")
+	if resp.Responses[2].GetResponseRange().Count > 0 {
+		err = proto.Unmarshal(resp.Responses[2].GetResponseRange().Kvs[0].Value, &rules)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to unmarshal rules")
+		}
+	} else if version != "" {
+		// if no rules are returned but a version is specified, it means that the version doesn't exist
+		return nil, api.ErrRulesetNotFound
 	}
+
+	// if the version wasn't specified, get the latest returned
 	if version == "" {
-		_, version = s.pathVersionFromKey(string(resp.Responses[2].GetResponseRange().Kvs[0].Key))
+		version = versions[len(versions)-1]
```

Comment:
We could also have stored the result of the `len` we did on the `kvs` above.  Micro-optimisation, mind you ;-)

Reply:
Actually, IIRC `len` is constant time, unlike C's `strlen` which is O(n) and justifies caching the length to reuse it.

Reply:
:+1: I knew it was constant time, but pushing it to the register, I didn't.